src/Functions/IFunction.cpp

```diff
@@ -102,13 +102,18 @@ ColumnPtr replaceLowCardinalityColumnsByNestedAndGetDictionaryIndexes(
     return indexes;
 }
 
-void convertLowCardinalityColumnsToFull(ColumnsWithTypeAndName & args)
+bool convertLowCardinalityColumnsToFull(ColumnsWithTypeAndName & args)
```

Comment:
The code in IFunction dealing with unpacking sparse / const / low-cardinality columns is already very complex and (before your PR) most humans will not be able to understand it. Imho, we should try to simplify it and not add more optimizations to make it even more complex.

Reply:
I found in `IExecutableFunction::executeWithoutSparseColumns`, the `use_count()` of input columns would increase  because of `ColumnsWithTypeAndName columns_without_low_cardinality = arguments;`, which prevents in-place optimizations in functions like `upper` and `lower`. That's the reason why I modified `IFunction.cpp`.

Reply:
For reasons of keeping the complexity down to a humanly reasonable level, I'd prefer to avoid changes in FunctionStringToString.h, and in IFunction.cpp (which are because of the changes in FunctionStringToString). Yes, that will not save as much runtime as before but in the long run, everyone will suffer from complex spaghetti code, thanks!

Reply:
I'm sorry that the optimization is impossible without workarounds in `IFunction.cpp`. Those workarounds  prevent input column's use_count from increasing for sparse/lowcardinality compatiability, so that we could apply in-place execution in lower/upper functions.